sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,745 @@
+            Y[feature_idx, sample_idx] = X[feature_idx, sample_idx]
+
+    return copyto_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_broadcast_division_kernel(n_samples, n_features, work_group_size):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+
+    @dpex.kernel
+    def broadcast_division(X, v):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        divisor = v[sample_idx]
+
+        for feature_idx in range(n_features):
+            X[feature_idx, sample_idx] = X[feature_idx, sample_idx] / divisor
+
+    return broadcast_division[global_size, work_group_size]
+
+
+@lru_cache
+def make_center_shift_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def center_shift(previous_center, center, center_shift):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        tmp = zero
+
+        for feature_idx in range(n_features):
+            center_diff = (
+                previous_center[feature_idx, sample_idx]
+                - center[feature_idx, sample_idx]
+            )
+            tmp += center_diff * center_diff
+
+        center_shift[sample_idx] = tmp
+
+    return center_shift[global_size, work_group_size]
+
+
+@lru_cache
+def make_half_l2_norm_dim0_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def half_l2_norm(X, result):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        l2_norm = zero
+
+        for feature_idx in range(n_features):
+            item = X[feature_idx, sample_idx]
+            l2_norm += item * item
+
+        result[sample_idx] = l2_norm / 2
+
+    return half_l2_norm[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_1dim_kernel(n_samples, work_group_size, device, dtype):
+    local_nb_iterations = math.floor(math.log2(work_group_size))
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group_id = dpex.get_group_id(0)
+        local_work_id = dpex.get_local_id(0)
+        first_work_id = local_work_id == 0
+
+        n_samples = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=dtype)
+
+        first_sample_idx = group_id * work_group_size * 2
+        augend_idx = first_sample_idx + local_work_id
+        addend_idx = first_sample_idx + work_group_size + local_work_id
+
+        if augend_idx >= n_samples:
+            shm[local_work_id] = zero
+        elif addend_idx >= n_samples:
+            shm[local_work_id] = v[augend_idx]
+        else:
+            shm[local_work_id] = v[augend_idx] + v[addend_idx]
+
+        dpex.barrier()
```

Comment:
I think I would find the code easier to understand if we always are explicit about which fence is used in a call to `barrier`:

- `dpex.CLK_LOCAL_MEM_FENCE` or
- `dpex.CLK_GLOBAL_MEM_FENCE`

Reply:
actually I tried to look-up what is the default of `numba_dpex.barrier` and it's not clear from the doc nor Python code interactive inspection (because the `barrier` method is just a stub that always error when interactively inspected outside of a kernel compilation context).

Reply:
From the doc
> Passing no arguments to numba_dpex.barrier() is equivalent to setting the global memory fence option.
so it's a major error from me, I thought the default was the opposite, all barriers are expected to be local in this PR.  :facepalm: